Library/Homebrew/formulary.rb

```diff
@@ -191,6 +193,27 @@ def self.from_rack(rack, spec=:stable)
   end
 
   def self.canonical_name(ref)
+    case ref
+    when %r{^([\w+-.]+)$}
+      return ref.downcase if (HOMEBREW_CELLAR/ref.downcase).directory?
+      if (link = (HOMEBREW_CELLAR/ref.downcase).resolved_path).directory?
+        basename = link.basename
+        if FORMULA_RENAMES[ref.downcase] == basename
+          return basename
+        end
+
+        if Tap.detect { |tap| tap.formula_renames[ref.downcase] == basename }
+          return basename
+        end
+      end
+    when HOMEBREW_TAP_FORMULA_REGEX
+      user, repo, name = ref.split("/", 3).map(&:downcase)
+      tap = Tap.new user, repo.sub(/^homebrew-/, "")
+      if (link = (HOMEBREW_CELLAR/name).resolved_path).directory?
+        return basename if tap.formula_renames[name] == link.basename
+      end
+    end
+
```

Comment:
Sorry, I forget to address that problem in the comment.

To solve your current problem, you should not change the `canonical_name` method. But you need to change the logic inside `ARGV#kegs` and a few similar logics in `uninstall`, `switch` etc. Instead using `canonical_name` directly to find the rack, you can do it in this way:

```ruby
canonical_name = Formulary.canonical_name(name)
rack = HOMEBREW_CELLAR/canonical_name

unless rack.directory?
  if name =~ HOMEBREW_TAP_FORMULA_REGEX
    rack = HOMEBREW_CELLAR/$3
  elsif !name.include("/“)
    rack = HOMEBREW_CELLAR/name
  end
end
```